services/kuber/server/nodes/delete.go

```diff
@@ -79,22 +81,42 @@ func (d *Deleter) DeleteNodes() (*pb.K8Scluster, error) {
 		}
 	}
 
+	// Cordon worker nodes to prevent any new pods/volume replicas being scheduled there
+	var eg errgroup.Group
 	for _, worker := range d.workerNodes {
-		// assure replication of storage
+		func(worker string) {
+			eg.Go(func() error {
+				return kubectl.KubectlCordon(worker)
+			})
+		}(worker)
+	}
+	if err := eg.Wait(); err != nil {
+		return nil, fmt.Errorf("error while cordoning worker nodes from cluster %s which were marked for deletion : %w", d.clusterPrefix, err)
+	}
+
+	// Remove worker nodes sequentially to minimise risk of fault when replicating PVC
+	for _, worker := range d.workerNodes {
+		// Assure replication of storage
 		if err := d.assureReplication(kubectl, worker); err != nil {
 			return nil, fmt.Errorf("error while making sure storage is replicated before deletion on cluster %s : %w", d.clusterPrefix, err)
 		}
-		// delete worker nodes from nodes.longhorn.io
+		// Delete worker nodes from nodes.longhorn.io
 		if err := d.deleteFromLonghorn(kubectl, worker); err != nil {
 			return nil, fmt.Errorf("error while deleting nodes.longhorn.io for %s : %w", d.clusterPrefix, err)
 		}
-		// delete worker nodes
+		// Delete worker nodes
 		if err := d.deleteNodesByName(kubectl, worker, realNodeNames); err != nil {
 			return nil, fmt.Errorf("error while deleting nodes from worker nodes for %s : %w", d.clusterPrefix, err)
 		}
+		// NOTE: Might need to verify if the volume got detached. The issue is still being discussed in #longhorn channel in CNCF slack workspace.
+		// https://cloud-native.slack.com/archives/CNVPEL9U3/p1683744899327089
+		// Issue seen, when once the node is deleted, some volumes might be still attached to the already deleted node, even though all replicas
+		// have moved to other nodes in cluster. Once manually detaching the volume from already deleted node (via UI), it attached again to a new node
+		// with no issues. Since there is no easy way how to verify and detach volume in some CLI way (due to longhorn policy of  doing everything via UI
+		// in ClickOps), leaving this note until discussion with longhorn maintainers concludes.
```

Comment:
Could we open a new Issue for this so we don't forget about this ?

Reply:
They just asked me to open an issue for it in the Longhorn repository. Should we open an issue on our end as well?

Reply:
Yes, and Link it with the Longhorn Issue

Reply:
https://github.com/berops/claudie/issues/784